src/main/java/com/yahoo/sketches/theta/HeapQuickSelectSketch.java

```diff
@@ -47,312 +47,363 @@
+		
+		
+		HashOperations.checkHashCorruption(hash);
+		empty_ = false;
+
+		
+		// The over-theta test
+		if (HashOperations.continueCondition(thetaLong_, hash)) {
+			return RejectedOverTheta; // signal that hash was rejected due to theta.
+		}
+
+		// The duplicate test
+		if (HashOperations.hashSearchOrInsert(cache_, lgArrLongs_, hash) >= 0) {
+			return RejectedDuplicate; // Duplicate, not inserted
+		}
+		
+		
+		// insertion occurred, must increment curCount
+		curCount_++;
+		
+
+		if (curCount_ > hashTableThreshold_) { // we need to do something, we are out of space
+			// must rebuild or resize
+			if (lgArrLongs_ <= lgNomLongs_) { // resize
+				resizeCache();
+			} else { // Already at tgt size, must rebuild
+			
+				assert (lgArrLongs_ == lgNomLongs_ + 1) : "lgArr: " + lgArrLongs_ + ", lgNom: " + lgNomLongs_;
+				quickSelectAndRebuild(); // Changes thetaLong_, curCount_, reassigns cache
+			}
+		}
+		return InsertedCountIncremented;
+
+	}
+
+	// Must resize. Changes lgArrLongs_ and cache_. theta and count don't change.
+	// Used by hashUpdate()
+	private final void resizeCache() {
+		final ResizeFactor rf = getResizeFactor();
+		final int lgTgtLongs = lgNomLongs_ + 1;
+		final int lgDeltaLongs = lgTgtLongs - lgArrLongs_;
+		final int lgResizeFactor = max(min(rf.lg(), lgDeltaLongs), 1); // rf_.lg() could be 0
+		lgArrLongs_ += lgResizeFactor; // new tgt size
+
+		final long[] tgtArr = new long[1 << lgArrLongs_];
+		final int newCount = HashOperations.hashArrayInsert(cache_, tgtArr, lgArrLongs_, thetaLong_);
+
+		assert newCount == curCount_; // Assumes no dirty values.
+		curCount_ = newCount;
+
+		cache_ = tgtArr;
+		hashTableThreshold_ = setHashTableThreshold(lgNomLongs_, lgArrLongs_);
+	}
+
+	// array stays the same size. Changes theta and thus count
+	private final void quickSelectAndRebuild() {
+		final int arrLongs = 1 << lgArrLongs_;
+
+		final int pivot = (1 << lgNomLongs_) + 1; // pivot for QS
+
+		thetaLong_ = selectExcludingZeros(cache_, curCount_, pivot); // messes up the cache_
+
+		// now we rebuild to clean up dirty data, update count, reconfigure as a hash
+		// table
+		final long[] tgtArr = new long[arrLongs];
+		curCount_ = HashOperations.hashArrayInsert(cache_, tgtArr, lgArrLongs_, thetaLong_);
+		cache_ = tgtArr;
+		// hashTableThreshold stays the same
+	}
+
+	/**
+	 * Returns the cardinality limit given the current size of the hash table array.
+	 *
+	 * @param lgNomLongs
+	 *            <a href="{@docRoot}/resources/dictionary.html#lgNomLongs">See
+	 *            lgNomLongs</a>.
+	 * @param lgArrLongs
+	 *            <a href="{@docRoot}/resources/dictionary.html#lgArrLongs">See
+	 *            lgArrLongs</a>.
+	 * @return the hash table threshold
+	 */
+	static final int setHashTableThreshold(final int lgNomLongs, final int lgArrLongs) {
+		final double fraction = (lgArrLongs <= lgNomLongs) ? RESIZE_THRESHOLD : REBUILD_THRESHOLD;
+		return (int) Math.floor(fraction * (1 << lgArrLongs));
+	}
+	
+	
+
+	
+
```

Comment:
So from this point on these are setters you added to use by the concurrent sub-classes?
If so then the methods can be protected and not public